### PR TITLE
[TIDY FIRST] Type `BaseRunner` class

### LINKED ISSUE #10606: [TIDY FIRST] Add typing to `BaseRunner` class
### Housekeeping

- [X] I am a maintainer of dbt-core

### Short description

The `BaseRunner` class is largely untyped. Given the prevalence of this class, not having it typed could mean that there are potential edge cases and/or runtime exceptions we are not aware of. By typing this class, we give ourselves a lot more guarantees

### Acceptance criteria

The `BaseRunner` class is fully typed

### Suggested Tests

N/A

### Impact to Other Teams

None

### Will backports be required?

No

### Context

I'm doing this as an example case for the Tech Improvement Tuesday we're having focused on typing improvements.

### PR DESCRIPTION
**Pull request by QMalcolm**
Resolves #10606

### Problem

The BaseRunner class is largely untyped. Given the prevalence of this class, not having it typed could mean that there are potential edge cases and/or runtime exceptions we are not aware of. By typing this class, we give ourselves a lot more guarantees

### Solution

Add typing. Of note, adding this typing uncovered a potential runtime exception in the `on_skip` method. We address this in [5adce38](https://github.com/dbt-labs/dbt-core/pull/10607/commits/5adce387a11424cb6a0930dc8b9a8eebc2d4d1a8).

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

File: .changes/unreleased/Under the Hood-20240826-141843.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+kind: Under the Hood
+body: Add type hinting to BaseRunner class to increase stability guarnatees
```
Comment (emmyoop) [inline .changes/unreleased/Under the Hood-20240826-141843.yaml:2]:
```suggestion
body: Add type hinting to BaseRunner class to increase stability guarantees
```

**Comment by QMalcolm** [comment]
The BaseRunner and it's derivatives unfortunately need to be refactored before we can properly type them 😞
